BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,190 @@
+            val plaintext =
+                cryptographyManager.decryptData(encryptedServerTokenWrapper.ciphertext, this)
+            SampleAppUser.fakeToken = plaintext
+            // now that you have the token, you can query server for everything else
+            // the only reason we call this fakeToken is because we didn't really get it from
+            // the server. In your case, you will have gotten it from the server the first time
+            // and therefore, it's a real token.
+
+            updateApp(getString(R.string.already_signedin))
+        }
+    }
+
+    // USERNAME + PASSWORD SECTION
+
+    private fun loginWithPasswordManager() {
+        loginWithPasswordViewModel.loginWithPasswordFormState.observe(this, Observer {
+            val loginState = it ?: return@Observer
+            when (loginState) {
+                is SuccessfulLoginFormState -> binding.login.isEnabled = loginState.isDataValid
+                is FailedLoginFormState -> {
+                    loginState.usernameError?.let { binding.username.error = getString(it) }
+                    loginState.passwordError?.let { binding.password.error = getString(it) }
+                }
+            }
+        })
+
+        binding.username.afterTextChanged {
+            loginWithPasswordViewModel.onLoginDataChanged(
+                binding.useBiometrics.text.toString(),
+                binding.password.text.toString()
+            )
+        }
+
+        binding.password.apply {
+            afterTextChanged {
+                loginWithPasswordViewModel.onLoginDataChanged(
+                    binding.username.text.toString(),
+                    binding.password.text.toString()
+                )
+            }
+
+            setOnEditorActionListener { _, actionId, _ ->
+                when (actionId) {
+                    EditorInfo.IME_ACTION_DONE ->
+                        loginWithPassword(
+                            binding.username.text.toString(),
+                            binding.password.text.toString()
+                        )
+                }
+                false
+            }
+        }
+        binding.login.setOnClickListener {
+            loginWithPassword(binding.username.text.toString(), binding.login.text.toString())
+        }
+
+        Log.d(TAG, "Username ${SampleAppUser.username}; fake token ${SampleAppUser.fakeToken}")
+    }
+
+    private fun loginWithPassword(username: String, password: String) {
+        val succeeded = loginWithPasswordViewModel.login(username, password)
+        if (succeeded) {
+            updateApp(
+                "You successfully signed up using password as: user " +
+                        "${SampleAppUser.username} with fake token ${SampleAppUser.fakeToken}"
+            )
+        }
+    }
+
+    private fun updateApp(successMsg: String) {
+        binding.success.text = successMsg
+    }
+}
+
+//Extension functions
+
+fun AppCompatEditText.afterTextChanged(task: () -> Unit) {
```

Comment:
Use [core-ktx](https://developer.android.com/kotlin/ktx/extensions-list#for_androidwidgettextview) instead.

Reply:
Nice, did realized they had one like that.